subsys/usb/device_next/class/usbd_hid.c

```diff
@@ -0,0 +1,743 @@
+	struct hid_device_data *ddata = dev->data;
+	struct usbd_hid_descriptor *desc = ddata->desc;
+
+	return desc->out_ep.bEndpointAddress;
+}
+
+static int usbd_hid_request(struct usbd_class_data *const c_data,
+			    struct net_buf *const buf, const int err)
+{
+	struct usbd_contex *uds_ctx = usbd_class_get_ctx(c_data);
+	const struct device *dev = usbd_class_get_private(c_data);
+	struct hid_device_data *ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	struct udc_buf_info *bi;
+
+	bi = udc_get_buf_info(buf);
+
+	if (bi->ep == hid_get_out_ep(c_data)) {
+		if (ops->output_report != NULL) {
+			if (err == 0) {
+				ops->output_report(dev, buf->len, buf->data);
+			}
+
+			k_work_submit(&ddata->output_work);
+		}
+	}
+
+	if (bi->ep == hid_get_in_ep(c_data)) {
+		if (ops->input_report_done != NULL) {
+			ops->input_report_done(dev);
+		} else {
+			k_sem_give(&ddata->in_sem);
+		}
+	}
+
+	return usbd_ep_buf_free(uds_ctx, buf);
+}
+
+static int handle_set_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup)
+{
+	const uint32_t duration = HID_GET_IDLE_DURATION(setup->wValue);
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		ddata->idle_rate = duration;
+	}
+
+	LOG_DBG("Set Idle, Report ID %u Duration %u", id, duration);
+
+	if (ops->set_idle != NULL) {
+		ops->set_idle(dev, id, duration * 4UL);
+	} else {
+		errno = -ENOTSUP;
+	}
+
+	return 0;
+}
+
+static int handle_get_idle(const struct device *dev,
+			   const struct usb_setup_packet *const setup,
+			   struct net_buf *const buf)
+{
+	const uint8_t id = HID_GET_IDLE_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+	uint32_t duration;
+
+	if (setup->wLength != 1U) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id != 0U && ops->get_idle == NULL) {
+		errno = -ENOTSUP;
+		return 0;
+	}
+
+	if (id == 0U) {
+		/* Only the common idle rate is stored. */
+		duration = ddata->idle_rate;
+	} else {
+		duration = ops->get_idle(dev, id) / 4UL;
+	}
+
+	LOG_DBG("Get Idle, Report ID %u Duration %u", id, duration);
+	net_buf_add_u8(buf, duration);
+
+	return 0;
+}
+
+static int handle_set_report(const struct device *dev,
+			     const struct usb_setup_packet *const setup,
+			     const struct net_buf *const buf)
+{
+	const uint8_t type = HID_GET_REPORT_TYPE(setup->wValue);
+	const uint8_t id = HID_GET_REPORT_ID(setup->wValue);
+	struct hid_device_data *const ddata = dev->data;
+	const struct hid_device_ops *ops = ddata->ops;
+
+	if (ops->set_report == NULL) {
+		errno = -ENOTSUP;
+		LOG_DBG("Set Report not supported");
+		return 0;
+	}
+
+	switch (type) {
+	case HID_REPORT_TYPE_INPUT:
+		LOG_DBG("Set Report, Input Report ID %u", id);
+		errno = ops->set_report(dev, type, id, buf->len, buf->data);
```

Comment:
Maybe we could make the `HID_REPORT_TYPE_` enum part of the application API and use it to pass report type?